sympy/integrals/transforms.py

```diff
@@ -1186,499 +1194,363 @@ def _laplace_build_rules(t, s):
     y = Wild('y')
     z = Wild('z')
     ma1 = f.match(exp(y)*z)
     if ma1:
         ma2 = ma1[y].collect(t).match(a*t)
         if ma2:
-            debug('_laplace_apply_rules match:')
+            debug('_laplace_apply_prog_rules match:')
             debug('      f:    %s ( %s, %s )'%(f, ma1, ma2))
-            debug('      rule: multiply with exp (1.5)')
-            L = _laplace_apply_rules(ma1[z], t, s-ma2[a], doit=doit, **hints)
-            noconds = hints.get('noconds', False)
-            if not noconds and type(L) is tuple:
-                r, p, c = L
-                return (r, p+ma2[a], c)
+            debug('      rule: multiply with exp (4.1.5)')
+            r, pr, cr = LaplaceTransform(ma1[z], t, s-ma2[a]).doit(noconds=False)
+            return True, (k*r, Max(p+re(ma2[a]), pr), And(c, cr))
+    return False, (fn, p, c)
+
+def _laplace_rule_delta(f, t, s):
+    """
+    If this function finds a factor ``DiracDelta(b*t-a)``, it applies the
+    masking property of the delta distribution. For example, if it gets
+    ``(DiracDelta(t-a)*f(t), t, s)``, it will return
+    ``(f(a)*exp(-a*s), -a, True)``.
+    """
+    # This rule is not in Bateman54
+    fn, p, c = _laplace_ct(f)
+    k, f = fn.as_independent(t, as_Add=False)
+
+    a = Wild('a', exclude=[t])
+    b = Wild('b', exclude=[t])
+
+    y = Wild('y')
+    z = Wild('z')
+    ma1 = f.match(DiracDelta(y)*z)
+    if ma1 and not ma1[z].has(DiracDelta):
+        ma2 = ma1[y].collect(t).match(b*t-a)
+        if ma2:
+            debug('_laplace_apply_prog_rules match:')
+            debug('      f:    %s ( %s, %s )'%(f, ma1, ma2))
+            debug('      rule: multiply with DiracDelta')
+            loc = ma2[a]/ma2[b]
+            if re(loc)>=0 and im(loc)==0:
+                r = exp(-ma2[a]/ma2[b]*s)*ma1[z].subs(t, ma2[a]/ma2[b])/ma2[b]
+                return True, (k*r, Max(p, -ma2[a]/ma2[b]), c)
             else:
-                return L
-    return None
+                return True, _laplace_ct(0)
+    return False, (fn, p, c)
 
 def _laplace_rule_trig(f, t, s, doit=True, **hints):
     """
-    This internal helper function tries to transform a product containing a
-    trigonometric function (`sin`, `cos`, `sinh`, `cosh`, ) and returns
-    `None` if it cannot do it.
+    This function covers trigonometric factors. All of the rules have a
+    similar form: ``trig(y)*z`` is matched, and then two copies of the Laplace
+    transform of `z` are shifted in the s Domain and added with a weight.
+
+    The parameters in the tuples are (fm, nu, s1, s2, sd):
+      fm: Function to match
+      nu: Number of the rule, for debug purposes
+      s1: weight of the sum, 'I' for sin and '1' for all others
+      s2: sign of the second copy of the Laplace transform of z
+      sd: shift direction; shift along real or imaginary axis if `1` or `I`
+
+    The convergence plane is changed only if the frequency shift is done along
+    the real axis.2
```

Comment:
Better: `if a frequency shift ...`; there is also a spurious `2` at the end of `axis.2`